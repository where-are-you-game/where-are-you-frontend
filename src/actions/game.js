import * as type from "../constants/actionTypes";

export const changeTextBox = (text) => ({
  type: type.CHANGE_TEXTBOX,
  text
});

export const changePlayerAnswer = (puzzleName, answer) => ({
  type: type.CHANGE_PLAYER_ANSWER,
  puzzleName,
  answer
});

export const changePlayerPassword = (passwordName, answer) => ({
  type: type.CHANGE_PLAYER_PASSWORD,
  passwordName,
  answer
});

export const solvePassword = (passwordName) => ({
  type: type.SOLVE_PASSWORD,
  passwordName
});

export const removePlayerGameData = () => ({
  type: type.REMOVE_PLAYER_GAME_DATA
});
