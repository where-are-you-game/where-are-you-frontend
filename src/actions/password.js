import * as type from "../constants/actionTypes";

export const savePasswords = (passwords) => ({
  type: type.SAVE_PASSWORDS,
  passwords
});
