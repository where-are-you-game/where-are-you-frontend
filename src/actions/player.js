import * as type from "../constants/actionTypes";

export const savePlayerName = (name) => ({
  type: type.SAVE_NAME,
  name
});

export const saveClearTime = (clearTime) => ({
  type: type.SAVE_NAME,
  clearTime
});

export const removePlayer = () => ({
  type: type.REMOVE_PLAYER
});

export const countClearTime = () => ({
  type: type.COUNT_CLEARTIME
});
