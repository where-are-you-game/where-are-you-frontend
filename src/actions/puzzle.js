import * as type from "../constants/actionTypes";

export const savePuzzles = (puzzles) => ({
  type: type.SAVE_PUZZLES,
  puzzles
});
