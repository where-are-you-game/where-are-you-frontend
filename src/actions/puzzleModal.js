import * as type from "../constants/actionTypes";

export const showPuzzleModal = (puzzle) => ({
  type: type.SHOW_PUZZLE_MODAL,
  puzzle
});

export const hidePuzzleModal = () => ({
  type: type.HIDE_PUZZLE_MODAL
});
