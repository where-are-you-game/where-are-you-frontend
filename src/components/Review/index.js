import React, { useState, useRef, useCallback } from "react";

import { useSelector, useDispatch } from "react-redux";
import styled from "styled-components";

import useFetchReviews from "../../hooks/useFetchReviews";
import LinkButton from "../Shared/LinkButton";
import View from "./View";
import Write from "./Write";

const Wrapper = styled.div`
  width: 800px;
`;

const TopBox = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

const Title = styled.p`
  margin: 0;
  font-size: 4rem;
  font-weight: 700;
`;

const Alert = styled.div`
  width: 100%;
  margin: 0 0 2.5rem 0;
  padding: ${({ theme }) => theme.padding.big};
  display: flex;
  align-items: center;
  justify-content: center;
  background: rgba(0, 0, 0, 0.3);
  color: #ffffff;
`;

const Notice = styled.div`
  margin: 0 0 0.3rem 0;
  color: ${({ theme }) => theme.color.red};
  font-size: 1rem;
`;

function Review() {
  const [page, setPage] = useState(1);
  const isLastPasswordUnlocked = useSelector((state) => state.game.playerPassword["password5"].isUnlocked);
  const hasReview = useSelector((state) => state.player.hasReview);
  const reviews = useSelector((state) => state.review.list);
  const observer = useRef();
  const dispatch = useDispatch();

  const {
    isLoading,
    error,
    hasMore
  } = useFetchReviews(page, dispatch);

  const lastReviewElementRef = useCallback((node) => {
    if (isLoading) return;

    if (observer.current) {
      observer.current.disconnect();
    }

    observer.current = new IntersectionObserver((entries) => {
      if (entries[0].isIntersecting && hasMore) {
        setPage((prev) => prev + 1);
      }
    });

    if (node) {
      observer.current.observe(node);
    }
  }, [isLoading, hasMore]);

  const renderWrite = () => {
    if (isLastPasswordUnlocked === false) {
      return <Alert>엔딩을 본 이후에 리뷰를 남길 수 있습니다.</Alert>;
    }

    if (hasReview) {
      return <Alert>리뷰는 한 번만 작성할 수 있습니다.</Alert>;
    }

    return <Write />;
  };

  const renderReviews = () => {
    return reviews.map((review, index) => {
      if (reviews.length === 0) {
        return <p>등록된 리뷰가 없습니다.</p>;
      }

      if (reviews.length === index + 1) {
        return (
          <View
            lastRef={lastReviewElementRef}
            key={review._id}
            name={review.name}
            clearTime={review.clearTime}
            content={review.content}
          />
        );
      }

      return (
        <View
          lastRef={null}
          key={review._id}
          name={review.name}
          clearTime={review.clearTime}
          content={review.content}
        />
      );
    });
  };

  return (
    <Wrapper>
      <TopBox>
        <Title>REVIEW</Title>
        <LinkButton
          path="/"
          title="메인으로 가기"
          color="black"
        />
      </TopBox>
      <Notice>* 게임에 대한 스포일러와 비밀번호는 피해서 적어주세요!!</Notice>
      {renderWrite()}
      {renderReviews()}
      {isLoading && <p>Loading...</p>}
      {error && <p>{error}</p>}
    </Wrapper>
  );
}

export default Review;
