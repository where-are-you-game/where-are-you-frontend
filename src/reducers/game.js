import produce from "immer";

import * as type from "../constants/actionTypes";

const makeAnswer = (name, count) => {
  const result = {};
  for (let i = 1; i <= count; i++) {
    result[`${name}${i}`] = {
      answer: ""
    };

    if (name === "password") {
      result[`${name}${i}`].isUnlocked = false;
    }
  }

  return result;
};

const initialState = {
  playerAnswer: makeAnswer("puzzle", 20),
  playerPassword: makeAnswer("password", 5),
  textBox: "테이블 위에 편지가 있다. 누가 쓴 거지? 읽어봐야겠다.",
  error: null
};

const game = (state = initialState, action) => {
  switch (action.type) {
    case type.CHANGE_TEXTBOX:
      return produce(state, (draft) => {
        draft.textBox = action.text;
      });
    case type.CHANGE_PLAYER_ANSWER:
      return produce(state, (draft) => {
        draft.playerAnswer[action.puzzleName].answer = action.answer;
      });
    case type.CHANGE_PLAYER_PASSWORD:
      return produce(state, (draft) => {
        draft.playerPassword[action.passwordName].answer = action.answer;
      });
    case type.SOLVE_PASSWORD:
      return produce(state, (draft) => {
        draft.playerPassword[action.passwordName].isUnlocked = true;
      });
    default:
      return state;
  }
};

export default game;
