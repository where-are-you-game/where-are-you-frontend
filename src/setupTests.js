import React from "react";

import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import ReactDOM from "react-dom";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
import { createStore } from "redux";
import { ThemeProvider } from "styled-components";

import { ModalProvider } from "./contexts/ModalContext";
import reducer from "./reducers";
import GlobalStyles from "./styles";
import GlobalFonts from "./styles/fonts";
import theme from "./styles/theme";

export const store = createStore(reducer);

export const history = createMemoryHistory({
  initialEntries: ["/"]
});

export const customRender = (children) => (
  render((
    <Router history={history}>
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <GlobalStyles />
          <GlobalFonts />
          <ModalProvider>
            {children}
          </ModalProvider>
        </ThemeProvider>
      </Provider>
    </Router>
  ))
);

beforeAll(() => {
  ReactDOM.createPortal = jest.fn((element) => element);
});
