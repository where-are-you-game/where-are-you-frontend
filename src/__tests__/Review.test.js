import React from "react";

import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { solvePassword } from "../actions/game";
import Review from "../components/Review";
import { customRender, store } from "../setupTests";

describe("Review", () => {
  it("matches snapshot", () => {
    const { container } = customRender(<Review />);

    expect(container).toMatchSnapshot();
  });

  it("should not write review if last password is locked", () => {
    customRender(<Review />);
    const notice = screen.getByText("엔딩을 본 이후에 리뷰를 남길 수 있습니다.");

    expect(notice).toBeInTheDocument();
  });
});

describe("write review", () => {
  it("should show validation text if textarea is empty", () => {
    store.dispatch(solvePassword("password5"));
    customRender(<Review />);
    const button = screen.getByRole("button", { name: "등록하기" });

    userEvent.click(button);

    const validationText = screen.getByText("내용을 입력해주세요.");

    expect(validationText).toBeInTheDocument();
  });

  it("should save review", () => {
    store.dispatch(solvePassword("password5"));
    customRender(<Review />);
    const textarea = screen.getByLabelText("리뷰 작성");
    const button = screen.getByText("등록하기");

    userEvent.type(textarea, "Hello, world!");
    userEvent.click(button);

    const review = screen.getByText("Hello, world!");

    expect(review).toBeInTheDocument();
  });
});
