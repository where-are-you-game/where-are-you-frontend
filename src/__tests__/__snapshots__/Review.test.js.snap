// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review matches snapshot 1`] = `
<div>
  <div
    class="sc-hBMUJo cWLnfk"
  >
    <div
      class="sc-fnVZcZ fCmaQd"
    >
      <p
        class="sc-fFSPTT evvNY"
      >
        REVIEW
      </p>
      <a
        class="sc-hKFxyN cEASwB"
        color="black"
        href="/"
      >
        <span>
          메인으로 가기
        </span>
      </a>
    </div>
    <div
      class="sc-iemWCZ fknPEp"
    >
      * 게임에 대한 스포일러와 비밀번호는 피해서 적어주세요!!
    </div>
    <div
      class="sc-bkbkJK fOPJOB"
      color="#ffffff"
    >
      엔딩을 본 이후에 리뷰를 남길 수 있습니다.
    </div>
    <div
      class="sc-bkbkJK cSmHcY"
      color="black"
    >
      등록된 리뷰가 없습니다.
    </div>
    <p>
      Loading...
    </p>
    <button
      class="sc-dIvrsQ iKdlMb"
      type="button"
    >
      <img
        alt="위로 가기 아이콘"
        src="top_icon.png"
      />
      TOP
    </button>
  </div>
</div>
`;
